ext/openssl/ossl_pkey_rsa.c

```diff
@@ -626,6 +626,174 @@ ossl_rsa_to_public_key(VALUE self)
+	goto err;
+
+    if (EVP_PKEY_CTX_set_rsa_padding(pkey_ctx, RSA_PKCS1_PSS_PADDING) != 1)
+	goto err;
+
+    if (EVP_PKEY_CTX_set_rsa_pss_saltlen(pkey_ctx, NUM2INT(saltlen)) != 1)
+	goto err;
+
+    if (EVP_PKEY_CTX_set_rsa_mgf1_md(pkey_ctx, mgf1md) != 1)
+	goto err;
+
+    if (EVP_DigestSignUpdate(md_ctx, RSTRING_PTR(data), RSTRING_LEN(data)) != 1)
+	goto err;
+
+    if (EVP_DigestSignFinal(md_ctx, (unsigned char *)RSTRING_PTR(signature), &buf_len) != 1)
+	goto err;
+
+    rb_str_set_len(signature, buf_len);
+
+    EVP_MD_CTX_free(md_ctx);
+    return signature;
+
+ err:
+    EVP_MD_CTX_free(md_ctx);
+    ossl_raise(eRSAError, NULL);
+}
+
+/*
+ *  call-seq:
+ *      pkey.verify_pss(digest, signature, data, salt_len, hash_alg) -> String
+ *
+ * Probabilistic Signature Scheme for RSA verify().
+ *
+ * To verify the +String+ +signature+, +digest+ and +data+ must be provided
+ * to re-compute the message digest. See +sign_pss+ for details about
+ * +digest+ and +data+.
+ *
+ * The +Integer+ +salt_len+ should be the salt length to use.
+ *
+ * The +String+ +hash_alg+ should be the hash algorithm used in MGF.
+ *
+ * The return value is +true+ if the
+ * signature is valid, +false+ otherwise. A PKeyError is raised should errors
+ * occur.
+ * Any previous state of the +Digest+ instance is irrelevant to the validation
+ * outcome, the digest instance is reset to its initial state during the
+ * operation.
+ *
+ * == Example
+ *   data = 'Sign me!'
+ *   digest = OpenSSL::Digest::SHA256.new
+ *   pkey = OpenSSL::PKey::RSA.new(2048)
+ *   signature = pkey.sign_pss(digest, data, 20, 'SHA256')
+ *   pub_key = pkey.public_key
+ *   puts pub_key.verify_pss(digest, signature, data, 20, 'SHA256') # => true
+ */
+static VALUE
+ossl_rsa_verify_pss(VALUE self, VALUE digest, VALUE signature, VALUE data, VALUE saltlen, VALUE hash_alg)
+{
+    EVP_PKEY *pkey;
+    EVP_PKEY_CTX *pkey_ctx;
+    const EVP_MD *md, *mgf1md;
+    EVP_MD_CTX *md_ctx;
+    int result;
+    RSA *rsa;
+    const BIGNUM *rsa_n;
+
+    GetPKeyRSA(self, pkey);
+    md = GetDigestPtr(digest);
+    mgf1md = GetDigestPtr(hash_alg);
+    StringValue(signature);
+    StringValue(data);
+
+    GetRSA(self, rsa);
+    RSA_get0_key(rsa, &rsa_n, NULL, NULL);
+    if (!rsa_n)
+	ossl_raise(eRSAError, "incomplete RSA");
+
+    md_ctx = EVP_MD_CTX_new();
+    if (!md_ctx)
+	ossl_raise(eRSAError, "EVP_MD_CTX_new");
+
+    if (EVP_DigestVerifyInit(md_ctx, &pkey_ctx, md, NULL, pkey) != 1)
+	goto err;
+
+    if (EVP_PKEY_CTX_set_rsa_padding(pkey_ctx, RSA_PKCS1_PSS_PADDING) != 1)
+	goto err;
+
+    if (EVP_PKEY_CTX_set_rsa_pss_saltlen(pkey_ctx, NUM2INT(saltlen)) != 1)
+	goto err;
+
+    if (EVP_PKEY_CTX_set_rsa_mgf1_md(pkey_ctx, mgf1md) != 1)
+	goto err;
+
+    if (EVP_DigestVerifyUpdate(md_ctx, RSTRING_PTR(data), RSTRING_LEN(data)) != 1)
+	goto err;
+
+    result = EVP_DigestVerifyFinal(md_ctx, (unsigned char *)RSTRING_PTR(signature), RSTRING_LENINT(signature));
```

Comment:
Calls for RSTRING_LENINT() must be made before allocating `md_ctx` because it may raise RangeError.